File: ios/MassiveRPG/Domain/Protocol/DataSource/FirebaseAuthDataSource.swift
//
//  FirebaseAuthDataSource.swift
//  MassiveRPG-Prod
//
//  Created by Onur Var on 8.04.2024.
//

import UIKit

protocol FirebaseAuthDataSource {
    // Sign in with Google using GoogleSignIn to Firebase
    func signIn(viewController: UIViewController) async throws -> User

    // Sign out from Firebase
    func signOut() throws
}

File: android/app/src/main/java/com/onurvar/massiverpg/domain/protocol/datasource/FirebaseAuthDataSource.kt
package com.onurvar.massiverpg.domain.protocol.datasource

import android.content.Intent
import android.content.IntentSender
import com.onurvar.massiverpg.domain.model.User

interface FirebaseAuthDataSource {

    // Begin the sign in process. It returns an IntentSender if the process is successful.
    suspend fun beginSignIn(): IntentSender

    // Signs the user in with the given Intent.
    suspend fun signIn(intent: Intent): User

    // Check if the user is signed in. It returns true if the user is signed in, false otherwise.
    fun checkSignedInUser(): Boolean

    // Signs the user out.
    suspend fun signOut()

    fun getUserId(): String?
}